packages/generator/src/unique-name-finder.ts

```diff
@@ -0,0 +1,146 @@
+import {
+  reservedObjectPrototypeKeywords,
+  reservedVdmKeywords
+} from './name-formatting-reserved-key-words';
+
+type Separator = '-' | '_';
+
+export class UniqueNameFinder {
+  private static readonly MAXIMUM_NUMBER_OF_SUFFIX = 1000;
+
+  public static getInstance(): UniqueNameFinder {
+    return new UniqueNameFinder();
+  }
+
+  private separator: Separator = '_';
+  private alreadyUsedNames: string[] = [];
+
+  /**
+   * Changes the default separator "_" to a different value.
+   * @param separator The separator to be used
+   * @returns The instance of the finder.
+   */
+  public withSeparator(separator: Separator): UniqueNameFinder {
+    this.separator = separator;
+    return this;
+  }
+
+  /**
+   * Sets the already used names considered in the finding process. Reserved TS keywords are always checked.
+   * @param names List of already used names
+   * @returns The instance of the finder.
+   */
+  public withAlreadyUsedNames(names: string[]) {
+    this.alreadyUsedNames = names.slice();
+    return this;
+  }
+
+  /**
+   * Adds the name(s) to the already used names.
+   * @param nameOrnames Names to be added
+   * @returns The instance of the finder.
+   */ public addToAlreadyUsedNames(nameOrnames: string | string[] | string) {
+    if (typeof nameOrnames === 'string') {
+      this.alreadyUsedNames.push(nameOrnames);
+    } else {
+      this.alreadyUsedNames.push(...nameOrnames);
+    }
+
+    return this;
+  }
+
+  /**
+   * Find a unique name by appending a suffix of the form this.separator\d+ if necessary. If the name is already unique nothing is appended.
+   * @param name The name to get a unique name from
+   * @returns Unique name
+   */
+  public findUniqueName(name: string): string {
+    return this.findUniqueNameWithSuffixes(name, [])[0];
+  }
+
+  /**
+   * Find a unique name by appending a suffix of the form this.separator\d+ if necessary. If the name is already unique nothing is appended.
+   * The suffixes are a list of strings appended to the name and these build names are also checked for uniqueness.
+   *
+   * @param name The name to get a unique name from
+   * @param suffixes Additional name suffixed to be considered
+   * @returns Unique names. The length of this array is one plus the number of suffixes provided. The first entry corresponds to the given name.
+   */
+  public findUniqueNameWithSuffixes(
+    name: string,
+    suffixes: string[]
+  ): string[] {
+    const relevantAlreadyUsedNames = this.removeUnnecessaryUsedNames(name);
+    if (
+      this.areNamesFree(
+        this.getAllNames(name, suffixes),
+        relevantAlreadyUsedNames
+      )
+    ) {
+      return [name, ...this.getAllNames(name, suffixes)];
+    }
+    const suffix = this.getUniqueNameUsingSuffix(
+      name,
+      relevantAlreadyUsedNames,
+      suffixes
+    );
+    return this.addSuffixes(name, suffix, suffixes);
+  }
+
+  private areNamesFree(names: string[], alreadyUsedNames: string[]): boolean {
```

Comment:
Up to now we have been referring to unique names and already used names, 'free' is just another word for 'unique' or 'notAlreadyUsed'. Consider `areNamesUnique` or even more exact `areAllNamesUnique` or turn it around and call it `areNamesAlreadyUsed` (in my opinion most intuitive and actually what is being checked in the function).

Reply:
true, I had the other name first but was also not happy. So let's go for the alreadyUsed one